File: 297. Serialize and Deserialize Binary Tree.cpp
class Codec {
public:
    string serialize(TreeNode* root) {
        if (!root) return "null-";
        string res;
        queue<TreeNode*> q;
        q.push(root);
        while (!q.empty()) {
            int size = q.size();
            while (size--) {
                TreeNode* node = q.front();
                q.pop();
                if (!node) {
                    res += "null+";
                } else {
                    res += to_string(node->val) + "+";
                    q.push(node->left);
                    q.push(node->right);
                }
            }
        }
        return res;
    }

    TreeNode* deserialize(string data) {
        if (data.empty()) return nullptr;
        vector<string> arr;
        string temp;
        for (char c : data) {
            if (c == '+') {
                arr.push_back(temp);
                temp.clear();
            } else {
                temp.push_back(c);
            }
        }
        arr.erase(remove_if(arr.begin(), arr.end(), [](const string& s){ return s.empty(); }), arr.end());
        if (arr.empty() || arr[0] == "null") return nullptr;
        TreeNode* root = new TreeNode(stoi(arr[0]));
        queue<TreeNode*> q;
        q.push(root);
        int ind = 1;
        while (ind < (int)arr.size()) {
            TreeNode* node = q.front();
            q.pop();
            if (ind < (int)arr.size()) {
                string left = arr[ind++];
                if (left != "null") {
                    TreeNode* leftNode = new TreeNode(stoi(left));
                    node->left = leftNode;
                    q.push(leftNode);
                }
            }
            if (ind < (int)arr.size()) {
                string right = arr[ind++];
                if (right != "null") {
                    TreeNode* rightNode = new TreeNode(stoi(right));
                    node->right = rightNode;
                    q.push(rightNode);
                }
            }
        }
        return root;
    }
};

File: 297. Serialize and Deserialize Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if(root == null) return "";
        ArrayList<Integer> arr = new ArrayList<>();
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        arr.add(root.val);
        while(!q.isEmpty()){
            TreeNode node = q.poll();
            if(node.left == null) arr.add(1001);
            else{
                arr.add(node.left.val);
                q.offer(node.left);
            }
            
            if(node.right == null) arr.add(1001);
            else{
                arr.add(node.right.val);
                q.offer(node.right);
            }
        }

        for(int num : arr) System.out.print(num + " ");
        StringBuilder sb = new StringBuilder();

        for(int i = 0; i < arr.size(); i++){
            sb.append(arr.get(i));
            sb.append("#");
        }

        return sb.toString();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if(data.length() == 0) return null;
        String[] vals = data.split("#");
        TreeNode root = new TreeNode(Integer.parseInt(vals[0]));
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        int index = 1;

        while(!q.isEmpty() && index < vals.length){
            TreeNode curr = q.poll();
            if(!vals[index].equals("1001")){
                curr.left = new TreeNode(Integer.parseInt(vals[index]));
                q.offer(curr.left);
            }

            index++;
            if(index < vals.length && !vals[index].equals("1001")){
                curr.right = new TreeNode(Integer.parseInt(vals[index]));
                q.offer(curr.right);
            }

            index++;
        }

        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));